Add rfc for k8s multi cluster deployment (#5069)
pipe-cd/pipecd: docs/rfcs/0014-multi-cluster-deployment-for-k8s.md

- Start Date: 2024-07-25
- Target Version: 0.49.0

# Summary

This RFC proposes a new feature for k8s app to deploy resources into multi-cluster.

# Motivation

# Usecase
- case 1. When applying the same manifest to multiple clusters for redundant configuration
- case 2. When applying manifest with some patches applied to multiple clusters for redundant configuration

> - case 3. Blue/Green Deployment acress the clusters

Reviewer: [nits] 
```suggestion
- case 3. Blue/Green Deployment across clusters
```

# Detailed design

## Overview

We propose the feature to apply the manifests to multiple-clusters in one application.

![image](assets/0014-pipeline-image.png)

## How it works

### Register Application with multiple platform providers

On the application register part, we can choose multiple platform providers.

> At first, user add the first platform provider. 

Reviewer: ```suggestion
At first, a user adds the first platform provider. 
```

If we want to use the feature for deploying multi-cluster, we can set more platform providers. This is optional.
Only the platform providers specified here can be configured for multi-target.

![image](assets/0014-choose-multiple-providers.png)

> Also, we can can the list of platform providers on the piped list page to verify the platform providers.

Reviewer: [nits] plz correct the verb. 
```suggestion
Also, we can xxx the list of platform providers on the piped list page to verify the platform providers.
```

    team: product
  quickSync:
    prune: true
    multiTarget:
      - provider:
          name: cluster-hoge
        resourceDir: ./cluster-hoge
      - provider:
          name: cluster-fuga
        resourceDir: ./cluster-fuga
  pipeline:
    stages:
      - name: K8S_CANARY_ROLLOUT
        with:
          replicas: 10%
          multiTarget:
            - provider:
                name: cluster-hoge
              resourceDir: ./cluster-hoge
            - provider:
                name: cluster-fuga
              resourceDir: ./cluster-fuga
...
```

**Rollback**

When rolling back, multiple environments are rolled back at the same time based on the information specified in `spec.quickSync.multiTarget`.
If at least one of the rollback processes succeeds, we consider the rollback successful.
This ensures that the rollback is executed for other environments even if one of the deployment environments is inaccessible.

#### Stages to be supported

We introduce the feature into the stages where changes are made to resources on the cluster.

- K8S_PRIMARY_ROLLOUT
- K8S_CANARY_ROLLOUT
- K8S_CANARY_CLEAN
- K8S_BASELINE_ROLLOUT
- K8S_BASELINE_CLEAN
- K8S_TRAFFIC_ROUTING

> ### How to check the stage progress of each platform providers in the deployment

Reviewer: ```suggestion
### How to check the stage progress of each platform provider in the deployment
```

        - /cluster-fuga
```

- When deploying
    - Sync all clusters corresponding to prd.

- When rolling back
    - Roll back in the all previous state.

### Pros & Cons

**Pros**

- Only one app setting is required.
- You can operate WaitApproval for all clusters in one place.

**Cons**

- Cannot support cases where you want to change the number of replicas for only some clusters.

# Idea: Create a stage to sync apps

![image](assets/0014-pipeline-sync-app-stage-01.png)

![image](assets/0014-pipeline-sync-app-stage-02.png)

### UX

- When registering an application
    - Prepare one app.pipecd.yaml as a root application with sync app stage.
    - Prepare manifests and app.pipecd.yaml for each clusters and & register on UI.
    - Dir structure

```
        - /prd
            - app.pipecd.yaml
            - /base
            - /cluster-hoge
                - app.pipecd.yaml
            - /cluster-fuga
                - app.pipecd.yaml
```

- When deploying
    - Sync all clusters corresponding to prd when triggering the root app.
    - If you want to sync clusters partially, sync them as the each application.

- When rolling back
    - Roll back in the all previous state.

> 
> - Rollback時

Reviewer: these lines should be removed??
```suggestion

```
